### ISSUE TITLE
Help: returning a different OneOf from within a Match

### ISSUE THREAD

**Issue by mattfrear** [mcintyre321/OneOf#36]
Hi Harry

I'm new to OneOf and am having a play with it at work, mainly using it as an Option type via `OneOf<T, None>`.

I have a couple of questions - 1. how do we avoid massive nesting within Matches when you need to do something and then something else and then something else and then something else?

2. How do I return a different OneOf<> from within a Match? My use case is this:
```csharp
private async Task<OneOf<string, None>> GetContent(string requestUri, CorrelationId correlationId)
{
    var response = await client.GetAsync(requestUri); // client is an HttpClient

    var content = response.Content == null ? string.Empty : await response.Content.ReadAsStringAsync();

    if (response.IsSuccessStatusCode)
    {
        return content;
    }

    return new None();
}

public async Task<OneOf<Address, None>> AddressSearch(string term, CorrelationId correlationId)
{
    var requestUri = AddressSearchUri
        .AddQueryParam("client_id", clientId)
        .AddQueryParam("term", term);

    var content = await GetContent(requestUri, correlationId);

    return content.Match(
        str => 
        {
            var addressSearchResponse = str.FromJson<AddressSearchResponse>();

            if (addressSearchResponse?.RetrieveLocationResponse?.BusinessResult?.Success != "Y")
            {
                return new None();
            }

            if (addressSearchResponse?.RetrieveLocationResponse?.TypeAheadSearch == null)
            {
                return new None();
            }

            return new Address( // this line fails to compile - "Cannot implicitly convert type Address to OneOf.Types.None"
                addressSearchResponse.RetrieveLocationResponse.TypeAheadSearch[0].Elid,
                addressSearchResponse.RetrieveLocationResponse.TypeAheadSearch[0].Label);
        },
        none => new None()
    );
}
```

**Comment by rtkelly13** [comment]
Because OneOf uses implicits to convert a T1 -> OneOf<T1, T2> this is fine inside a normal function because the return type is explicit. 
But inside a lambda function this doesn't have to be the case. The result of the lambda can help infer the types on the method being used.

How does it know you want to return a Lambda, best way I have found is to annotate the Match method with the result type. Unfortunately it means writing the type twice! So Match<OneOf<T1, T2>> this means the Func given has to return a OneOf type and implicits work again.

**Comment by mcintyre321** [comment]
@rtkelly13 is correct, use `.Match<OneOf<Address, None>>(...)`

I'm just thinking that it might be possible to change the MapTX methods from 

`public OneOf<TResult, T1> MapT0<TResult>(Func<T0, TResult> mapFunc)`

to 

`public OneOf<TResult, T1> MapT0<TResult>(Func<T0, OneOf<TResult, T1>> mapFunc)`

and then you could use those methods.

**Comment by rtkelly13** [comment]
@mcintyre321 What you suggested is more of a classic Bind method rather than a Map method

https://fsharpforfunandprofit.com/series/map-and-bind-and-apply-oh-my.html

It could be a valuable addition and would simplify this case massively, Has more of the same problems as the Either monad being that BindT0/BindLeft (Or was that BindRight 😉) means nothing when reading it you have to know the type is OneOf<T, None> to figure out what the usage is for.

I could create a PR with the change if you are interested in its addition?

**Comment by mcintyre321** [comment]
Sound good...

...but I tried a little spike and the approach I took ended up using creating a LOT of overloads on each OneOf<T0, .., TN>

e.g. on `OneOf<T0, T1, T2>`

`public OneOf<T1, T2> BindT0(Func<T0, T1> mapFunc)`
`public OneOf<T1, T2> BindT0(Func<T0, T2> mapFunc)`
`public OneOf<T0, T2> BindT1(Func<T0, T0> mapFunc)`
`public OneOf<T0, T2> BindT1(Func<T0, T2> mapFunc)`
`public OneOf<T0, T1> BindT2(Func<T0, T0> mapFunc)`
`public OneOf<T0, T1> BindT2(Func<T0, T1> mapFunc)`

I'm a bit worried that it will balloon the assembly size, [which has been a problem in the past](https://github.com/mcintyre321/OneOf/issues/24)! 

If adding the new methods doesn't significantly increase the dll size, then, the more the merrier!

I'm wondering if Bind (and Map) could be moved into extension methods, in another assembly, I think they are the ones which baloon the assembly.

**Comment by mcintyre321** [comment]
Just read your username properly, hi Matt! Hope the move to NZ has gone well!

**Comment by mcintyre321** [comment]
> how do we avoid massive nesting within Matches when you need to do something and then something else and then something else and then something else?

I'm not sure there is a good way (you'd have the same situation with switches and ifs, right?) but extracting and local functions instead of the lambdas can make things a lot clearer.

**Comment by rtkelly13** [comment]
Local functions can be the nicer way to define the inner methods and reduce the type inference burden a lot! I will give the bind method a go at some point you were on the right path with your edited version, I think my original implementation was with extension methods but you were right to move it into the core OneOf type.

Moving those out into extension methods would also be a binary incompatible breaking change, which was an issue when you broke OneOf 8..32 out any libraries depending on you broke until they moved the base version above that release. Maybe a major release would be required in that case.

There are also type inference changes again which are subtle but hinder the API a lot, I will get an example together. 

In regards to assembly size the Map functions did multiply at a rate of https://en.m.wikipedia.org/wiki/1_%2B_2_%2B_3_%2B_4_%2B_⋯
This change would be the same number of items again, so a linear increase rather than an exponential one as before. Maybe a size comparison on the PR would be useful.

**Comment by mattfrear** [comment]
> 
> 
> > how do we avoid massive nesting within Matches when you need to do something and then something else and then something else and then something else?
> 
> I'm not sure there is a good way (you'd have the same situation with switches and ifs, right?) but extracting and local functions instead of the lambdas can make things a lot clearer.

With ifs I can exit the flow early to prevent nesting, i.e.
```
var result = doSomething(foo);
if (!result) return;

var otherResult = doOtherThing(result);

if (!otherResult.Success) return;

var anotherResult = doAnotherThing(otherResult);
```

I'm guessing to do something similar with OneOf would look like
```
OneOf<Result, None> result = doSomething(foo);
if (result.IsT1) return;

var otherResult = doOtherThing(result);
...
```
But this is pretty stinky... if I could do
```
OneOf<Result, None> result = doSomething(foo);
if (result is None) return;
```
that might be more what I'm used to. But I'm still new to all this and maybe that pattern is to be avoided?

PS. Thanks Harry, yep the move to NZ has gone well. Only took a week to find a contract 😃 Not quite London rates and the cost of living is noticeably higher here, but life is good so far. Are you still working at RP?

**Comment by mattfrear** [comment]
> 
> 
> Because OneOf uses implicits to convert a `T1` -> `OneOf<T1, T2>` this is fine inside a normal function because the return type is explicit.
> But inside a lambda function this doesn't have to be the case. The result of the lambda can help infer the types on the method being used.
> 
> How does it know you want to return a OneOf inside a lambda, best way I have found is to annotate the Match method with the result type. Unfortunately, it means writing the type twice! So `Match<OneOf<T1, T2>>` this means the Func given has to return a `OneOf` type and implicits work again.
> 
> TLDR; For your example changing `Match` - > `Match<OneOf<Address, None>>` will do the trick

Thanks @rtkelly13 👍

**Comment by rtkelly13** [comment]
I would say early return would not be particularly idiomatic in a functional language if we want to compare this to F#. Even though there are some parts of the language where multiple return statements are used.
https://fsharpforfunandprofit.com/series/computationexpressions.html

If you try to use the the IsTX methods you loose most of the type safety benefits of this library. They shouldn't really be used for purposes other than extensions such as Serialisation. They should be hidden somehow in my option as them being there encourages people to miss-use this library. 

My guidelines would be to be when you are creating a OneOf type use multiple return statements, but when you are consuming them you should use only one return statements. This forces the consumer to handle all the cases inside the OneOf! Avoiding the pitfalls of the is/as or multiple if statement approach!

**Comment by rtkelly13** [comment]
Created an example Dotnet Fiddle with a simple example of how a bind method would look when created as an extension method.

https://dotnetfiddle.net/nsAqLv

This is the issue when introducing Tasks too, you have to specify all the types on the OneOf and the result type if you can't infer all types they have to be provided explicitly.

**Comment by mcintyre321** [comment]
@mattfrear last day at RP today actually! Glad to hear you hit the ground running.

You might find the `.TryPickTX` methods are what you are after. This is from the readme:

```
OneOf<Thing, NotFound, Error> thingOrNotFoundOrError = GetThingFromDb(string id);

	if (thingOrNotFoundOrError.TryPickT1(out NotFound notFound, out var thingOrError)) //thingOrError is a OneOf<Thing, Error>
		return StatusCode(404);

	if (thingOrError.TryPickT1(out var error, out var thing)) //note that thing is a Thing rather than a OneOf<Thing>
	{
		_logger.LogError(error.Message);
		return StatusCode(500);
	}

	return Ok(thing);
```

This lets you remove a single case at a time, and process the OneOf without deep nesting
